File: pyalgopractice/leap_year.py
def check_leap_year(year: int) -> bool:
    """Function to determine if an input year is a leap year

    Args:
        int (year): Year, using the gregorian calendar

    Returns:
        bool: returns true if year is a leap year, and false if not
    """
    leap_bool = False

    if even_div(year, 4):
        leap_bool = True
        if even_div(year, 100):
            leap_bool = False
            if even_div(year, 400):
                leap_bool = True
            else:
                leap_bool = False
        else:
            leap_bool = True
    else:
        leap_bool = False
    return leap_bool


def even_div(year: int, stage: int) -> bool:
    """Function to check if a number is even

    Args:
        int (year): The number to determine if even
        int (stage): The number to be divided by

    Returns:
        bool: True if divisible with no remainder and False otherwise
    """
    if year % stage == 0:
        return True
    else:
        return False


if __name__ == "__main__":
    test_cases = [2000, 2004, 2012, 2023, 2012, 1024, 2005]

    for x in test_cases:
        if check_leap_year(x):
            print("It's a leap year")
        else:
            print("not a leap year")

    print("Finished Running")

File: calgopractice/leap_year.cpp
#include <iostream>

bool is_even(const int &year, const int &stage)
{
    /*
        This function returns true if the year is divisible without a remainder,
        and False if otherwise

        Parameters:
            int (year): year from the gregorian calendar
            int(stage): the division stages for lack of better wording
        Returns:
            bool: true or false depending on evenly divisible or not
    */
    if (year % stage == 0)
    {
        return true;
    }
    else
    {
        return false;
    }
}

bool is_leap_year(const int &year)
{
    /*
        Function returns true or false depending on
        if the input is a leap year

        Parameter:
            int(year): year from the gregorian calendar
        Returns:
            bool: true or false depending on leap year or not
    */
    bool leap_year = false;
    int stages[3] = {4, 100, 400};

    if (is_even(year, stages[0]))
    {
        leap_year = true;
        if (is_even(year, stages[1]))
        {
            leap_year = false;
            if (is_even(year, stages[2]))
            {
                leap_year = true;
            }
            else
            {
                leap_year = false;
            }
        }
    }
    else
    {
        leap_year = false;
    }
    return leap_year;
}

int main()
{

    int test_cases[8] = {2000, 2004, 2012, 2023, 2012, 1024, 2005}; // Instantiating test cases

    for (auto i : test_cases)
    {
        if (is_leap_year(i))
            std::cout << "This is a leap year \n ";
        else
            std::cout << "This is not a leap year \n";
    }

    return 0;
}
